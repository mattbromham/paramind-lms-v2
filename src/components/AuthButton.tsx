import { useState } from 'react';

import { Button } from '@/components/ui/button';
import { getSupabase } from '@/lib/supabase';
import { useSession } from '@/providers/AuthProvider';

import SignInDialog from './SignInDialog';

export default function AuthButton() {
  const supabase = getSupabase();
  const session = useSession();
  const [signInOpen, setSignInOpen] = useState(false);

  const handleSignOut = async () => {
    await supabase.auth.signOut();
  };

  if (session) {
    return (
      <div className="flex items-center gap-2">
        <Button variant="ghost" onClick={handleSignOut} aria-label="Sign out">
          Sign out
        </Button>
        <span className="text-sm text-pm-text-low ml-2">
          ({session.user.email})
        </span>
      </div>
    );
  }

  return (
    <>
      <Button
        variant="default"
        className="bg-pm-primary text-white"
        onClick={() => setSignInOpen(true)}
        aria-haspopup="dialog"
      >
        Sign in
      </Button>
      <SignInDialog open={signInOpen} onOpenChange={setSignInOpen} />
    </>
  );
}
