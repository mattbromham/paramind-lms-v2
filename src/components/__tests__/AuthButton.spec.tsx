import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import AuthButton from '../AuthButton';

// Mock the dependencies
vi.mock('@/lib/supabase', () => ({
  getSupabase: vi.fn(() => ({
    auth: {
      signOut: vi.fn().mockResolvedValue({}),
    },
  })),
}));

vi.mock('@/providers/AuthProvider', () => ({
  useSession: vi.fn(),
}));

vi.mock('../SignInDialog', () => ({
  default: ({
    open,
    onOpenChange: _onOpenChange,
  }: {
    open: boolean;
    onOpenChange: (open: boolean) => void;
  }) => <div data-testid="sign-in-dialog" data-open={open} />,
}));

describe('AuthButton', () => {
  const mockSignOut = vi.fn();
  const mockUseSession = vi.fn();

  beforeEach(async () => {
    vi.clearAllMocks();

    // Mock useSession
    const { useSession } = await import('@/providers/AuthProvider');
    vi.mocked(useSession).mockImplementation(mockUseSession);

    // Mock getSupabase
    const { getSupabase } = await import('@/lib/supabase');
    vi.mocked(getSupabase).mockReturnValue({
      auth: {
        signOut: mockSignOut,
      },
    } as unknown as ReturnType<typeof getSupabase>);
  });

  it('renders Sign In button when not authenticated', () => {
    mockUseSession.mockReturnValue(null);

    render(<AuthButton />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    expect(signInButton).toBeInTheDocument();
    expect(signInButton).toHaveClass('bg-pm-primary');
  });

  it('renders Sign Out button and email when authenticated', () => {
    const mockSession = {
      user: { email: 'test@example.com' },
    };
    mockUseSession.mockReturnValue(mockSession);

    render(<AuthButton />);

    const signOutButton = screen.getByRole('button', { name: /sign out/i });
    expect(signOutButton).toBeInTheDocument();
    expect(screen.getByText('(test@example.com)')).toBeInTheDocument();
  });

  it('calls signOut when Sign Out button is clicked', async () => {
    const user = userEvent.setup();
    const mockSession = {
      user: { email: 'test@example.com' },
    };
    mockUseSession.mockReturnValue(mockSession);

    render(<AuthButton />);

    const signOutButton = screen.getByRole('button', { name: /sign out/i });
    await user.click(signOutButton);

    expect(mockSignOut).toHaveBeenCalledOnce();
  });

  it('opens SignInDialog when Sign In button is clicked', async () => {
    const user = userEvent.setup();
    mockUseSession.mockReturnValue(null);

    render(<AuthButton />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    await user.click(signInButton);

    const dialog = screen.getByTestId('sign-in-dialog');
    expect(dialog).toHaveAttribute('data-open', 'true');
  });

  it('matches snapshot when logged out', () => {
    mockUseSession.mockReturnValue(null);

    const { container } = render(<AuthButton />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot when logged in', () => {
    const mockSession = {
      user: { email: 'test@example.com' },
    };
    mockUseSession.mockReturnValue(mockSession);

    const { container } = render(<AuthButton />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
