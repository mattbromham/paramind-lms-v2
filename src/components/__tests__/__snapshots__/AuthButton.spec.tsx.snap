// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthButton > matches snapshot when logged in 1`] = `
<div
  class="flex items-center gap-2"
>
  <button
    aria-label="Sign out"
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Sign out
  </button>
  <span
    class="text-sm text-pm-text-low ml-2"
  >
    (
    test@example.com
    )
  </span>
</div>
`;

exports[`AuthButton > matches snapshot when logged out 1`] = `
<button
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 bg-pm-primary text-white"
  data-slot="button"
>
  Sign in
</button>
`;
